### PR TITLE
Upgrading build process

### PR DESCRIPTION
**Pull request by MatApple**
## Feature branch
- No more virtual machine proxy for dcos ui. Proxy is part of dev server.
- Now have hot loading on dev server.
- No hard coded references to external plugins. External plugins location is now
completely configurable with npm config key. Now have full flexibility of where plugins are in relation to dcos-ui and the plugin directory name.
- No more copying over plugins into dcos-ui repo via gulp. Webpack will load and watch plugins specified in the directory set by 
```
npm config set dcos-ui:external_plugins ../relative/path/to/plugins/folder
```
- Use webpack's watch for all files. No more gulp watch.
- Full cache busting on all files. Every file (images, css imports e.g. icons, css, Js) has webpack's build hash prepended to it's file name. [hash]-name.[ext].
- Reduced built bundle to only 1.54MB (1.7MB with plugins) of JavaScript and 400kB of css. Reduced dev bundle size from 11MB of JavaScript to 7MB. Force React to build with NODE_ENV=production to reduce footprint of React. Pruned fixtures from production build.
- Dev build process no longer includes long running tasks like image optimization.
- Everything is now linted including unit and integration tests.
- Linting runs on plugins and dcos-ui on every file change and results show in terminal.
- Lint errors now force CI testing to fail early.
- Styles (css, less, sass) and images (png, jpeg, svg, etc) can now be bundled within plugins easily and imported as regular modules `import avatar from '../img/avatar';`

### PRs open (select once merged):
- [x] https://github.com/dcos/dcos-ui/pull/51
- [x] ~~https://github.com/dcos/dcos-ui/pull/52~~ [closed]
- [x] https://github.com/dcos/dcos-ui/pull/87
- [x] https://github.com/dcos/dcos-ui/pull/53
- [x] https://github.com/dcos/dcos-ui/pull/54

### NOTE:
None of the tests will pass on the above PRs till they all merge into this branch. I split them apart for easier code review. Don't bother testing locally till all code is in this PR.

### Outstanding:
- [x] Include Dygraphs in new build
- [x] Deal with Config.dev.js on production builds.
- [x] Remove all Gulpfile tasks except `connect:server` so we can still validate builds. [edit: Also use it for ensuring Config.dev.js is present on production builds.]
- [x] Remove gulp related dependencies from package.json
- [x] Shrinkwrap
- [x] Update README

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @philipnrmn

**Comment by jfurrow** [comment]
Nice work!

File: README.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 Node 4.x (and above) is **required**. We suggest using [nvm](https://github.com/creationix/nvm) or [n](https://github.com/tj/n) to keep multiple Node versions on your system.
 
 
+
```
Comment (kennyt) [inline README.md]: can you write unit tests, integration tests, docs, and create a wiki, and send out an email to the team for this change?

Comment (MatApple) [inline README.md]: 👍  It's a placeholder for bigger changes. The README will be updated with actual information.

Comment (kennyt) [inline README.md]: nah dude it was a joke cause it's just a newline

Comment (mlunoe) [inline README.md]: It's really hard to review this change. Could we break it into separate PRs?

Comment (jfurrow) [inline README.md]: Could you maybe add some comments that explains this change? Nothing huge, but two or three paragraphs about what's going on should be sufficient!

Comment (daiyiwhale) [inline README.md]: Can you test in all browsers? not sure if Edge is supported

File: webpack/webpack.dev.babel.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+module.exports = Object.assign({}, webpackConfig, {
+  entry: [
+    'webpack-dev-server/client?http://localhost:' + PORT,
+    'webpack/hot/only-dev-server',
+    './src/js/index.js'
+  ],
+  devtool: '#eval-source-map',
+  output: {
+    path: './build',
+    filename: 'index.js'
```
Comment (rcorral) [inline webpack/webpack.dev.babel.js]: Why are these paths hardcoded? Shouldn't they come from the config?

File: jest/genConfig.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
+/*
+  Generates configuration for jest.
+ */
+var fs = require('fs');
+var path = require('path');
+var pluginsFolder = process.env.DCOS_UI_PLUGINS;
+var testPaths = ['src'];
+if (pluginsFolder) {
+  try {
+    require('../' + pluginsFolder);
```
Comment (rcorral) [inline jest/genConfig.js]: What if the plugins path is an absolute path? won't this fail?

Comment (rcorral) [inline jest/genConfig.js]: Do we even need this require?

File: package.json [hunk 4]
```diff
@@ -47,121 +48,70 @@
...
+    "less": "2.6.1",
+    "less-loader": "2.2.3",
     "localStorage": "1.0.3",
     "marked": "0.3.5",
-    "mkdirp": "0.5.1",
     "node-libs-browser": "1.0.0",
-    "run-sequence": "1.1.5",
+    "react-hot-loader": "1.3.0",
+    "postcss-loader": "0.8.2",
+    "react-hot-loader": "1.3.0",
```
Comment (rcorral) [inline package.json]: this is here twice.
Also would've loved to see the added features on a different PR (post this massive one)

File: webpack/build.config.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+var files = {
+  srcJS: dirs.srcJS + '/index.js',
+  distJS: dirs.distJS + '/index.js',
+  srcCSS: dirs.srcCSS + '/index.less',
+  distCSS: dirs.distCSS + '/index.css',
+  srcHTML: dirs.src + '/index.html',
+  distHTML: dirs.dist + '/index.html'
+};
```
Comment (rcorral) [inline webpack/build.config.js]: It doesn't look like you use any of these variables.

File: package.json [hunk 6]
```diff
@@ -47,121 +48,70 @@
...
+    "test": "npm run lint && npm run lint-plugins && node jest/genConfig.js && NODE_PATH=node_modules jest --config=jest/config.json --no-cache",
+    "jest": "node jest/genConfig.js && NODE_PATH=node_modules jest --config=jest/config.json --no-cache",
+    "start": "webpack-dev-server --config ./webpack/webpack.dev.babel.js --hot --progress --colors --content-base ./build",
+    "build": "./node_modules/.bin/gulp ensureConfig && NODE_ENV='production' webpack --config ./webpack/webpack.production.babel.js"
   },
-  "jest-webpack-alias": {
-    "configFile": ".webpack.config.js"
+  "analytics": {
+    "development": "39uhSEOoRHMw6cMR6st9tYXDbAL3JSaP",
+    "production": "51ybGTeFEFU1xo6u10XMDrr6kATFyRyh"
```
Comment (rcorral) [inline package.json]: Not sure if this is the best spot for these config keys.

File: webpack/webpack.config.babel.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import autoprefixer from 'autoprefixer';
+import colorLighten from 'less-color-lighten';
+import path from 'path';
+
+function absPath() {
+  let args = [].slice.apply(arguments);
+  args.unshift(__dirname, '..');
+
+  return path.resolve.apply(path.resolve, args);
+}
```
Comment (rcorral) [inline webpack/webpack.config.babel.js]: Could you use this instead?
https://nodejs.org/api/path.html#path_path_normalize_path

File: webpack/webpack.dev.babel.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import ExtractTextPlugin from 'extract-text-webpack-plugin';
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import StringReplacePlugin from 'string-replace-webpack-plugin';
+import WebpackNotifierPlugin from 'webpack-notifier';
+
+import packageInfo from '../package';
+import webpackConfig from './webpack.config.babel';
+
+let PORT = process.env.PORT || 8080;
```
Comment (rcorral) [inline webpack/webpack.dev.babel.js]: Could we default to 4200 as that's what most people use today?

File: jest/genConfig.js (hunk 3 above)
Comment (MatApple) [inline jest/genConfig.js]: We need to check the path is valid. If it isn't a valid path then we don't add the path to jest because jest won't be able to resolve it. The next line will never execute.

File: webpack/webpack.dev.babel.js (hunk 2 above)
Comment (MatApple) [inline webpack/webpack.dev.babel.js]: What do you think about getting rid of the `build.config.js` file? Those variables aren't really needed because we only use them here. This webpack config is for this repo and doesn't need to be generalized.

File: package.json (hunk 4 above)
Comment (MatApple) [inline package.json]: Good catch.

File: webpack/build.config.js (hunk 5 above)
Comment (MatApple) [inline webpack/build.config.js]: You're right, I'd say we get rid of this file because it is now an unnecessary layer of abstraction.

File: package.json (hunk 6 above)
Comment (MatApple) [inline package.json]: Ideally they are part of the telemetry plugin. I'll have to check if we can put them there now that we've changed the way we do telemetry. Previously it was difficult because we had multiple dependents.

File: webpack/webpack.config.babel.js (hunk 7 above)
Comment (MatApple) [inline webpack/webpack.config.babel.js]: Sure!

File: webpack/webpack.dev.babel.js (hunk 8 above)
Comment (MatApple) [inline webpack/webpack.dev.babel.js]: Yes, makes sense!

File: webpack/webpack.dev.babel.js (hunk 2 above)
Comment (MatApple) [inline webpack/webpack.dev.babel.js]: I've removed `build.config.js`.

File: package.json (hunk 4 above)
Comment (MatApple) [inline package.json]: Fixed

File: webpack/build.config.js (hunk 5 above)
Comment (MatApple) [inline webpack/build.config.js]: Gone!

File: webpack/webpack.dev.babel.js (hunk 8 above)
Comment (MatApple) [inline webpack/webpack.dev.babel.js]: Done

File: package.json (hunk 6 above)
Comment (MatApple) [inline package.json]: Moved to Config file. There was only a single instance where we were replacing `@@ANALYTICS_KEY`.

File: webpack/webpack.dev.babel.js (hunk 8 above)
Comment (mlunoe) [inline webpack/webpack.dev.babel.js]: Did you push?

File: gulpfile.js [hunk 9]
```diff
@@ -1,259 +1,18 @@
...
-      baseDir: config.dirs.dist
-    },
-    socket: {
-      domain: 'localhost:4200'
-    }
-  });
 });
 
-gulp.task('connect:server', function () {
+gulp.task('serve', function () {
```
Comment (mlunoe) [inline gulpfile.js]: Won't this still be rivalling the webpack-dev-server?

File: gulpfile.js [hunk 10]
```diff
@@ -1,259 +1,18 @@
...
-}
 
-function ensurePluginDirectoryExists() {
-  // Make temp plugins directory if doesn't exist
-  if (!fs.existsSync(config.dirs.pluginsTmp)) {
-    mkdirp(config.dirs.pluginsTmp, function () {
-      console.log('Created ' + config.dirs.pluginsTmp);
-    });
+gulp.task('ensureConfig', function () {
+  // Make sure we have a Config.dev so we don't error on Config loading
```
Comment (mlunoe) [inline gulpfile.js]: Couldn't we do something similar to what we are doing with the proxy here?

File: package.json [hunk 11]
```diff
@@ -47,121 +48,70 @@
...
     "source-map-loader": "0.1.5",
+    "string-replace-webpack-plugin": "0.0.3",
+    "style-loader": "0.13.1",
     "transform-loader": "0.2.3",
     "webpack": "1.13.0",
+    "webpack-dev-server": "1.14.1",
     "webpack-notifier": "1.3.0"
   },
   "scripts": {
     "clean": "rm -rf build",
```
Comment (mlunoe) [inline package.json]: This has been a mistake for a long time, but the folder being produced is called 'dist': https://github.com/dcos/dcos-ui/pull/55/files#diff-4c60f7b3842aaf35cce1238f92a3ea6fR2

Comment (mlunoe) [inline package.json]: Not something we need to fix here

File: gulpfile.js [hunk 12]
```diff
@@ -1,261 +1,18 @@
...
-      domain: 'localhost:4200'
-    }
-  });
 });
 
-gulp.task('connect:server', function () {
+gulp.task('serve', function () {
   connect.server({
     port: 4200,
     root: config.dirs.dist
```
Comment (mlunoe) [inline gulpfile.js]: I think this file was just deleted

Comment (MatApple) [inline gulpfile.js]: Seems to still be alive!

File: package.json (hunk 11 above)
Comment (MatApple) [inline package.json]: Yeah that clean can just get removed. It's not cleaning anything!

File: gulpfile.js (hunk 10 above)
Comment (MatApple) [inline gulpfile.js]: This is a bit ugly but it works for now. Lets mark it as an improvement to make in the future. What do you think?

File: gulpfile.js (hunk 12 above)
Comment (mlunoe) [inline gulpfile.js]: Not for me

File: gulpfile.js (hunk 9 above)
Comment (MatApple) [inline gulpfile.js]: Yes, in terms of default port usage. Just don't run them at the same time, or change the port for dev server.

File: package.json (hunk 11 above)
Comment (mlunoe) [inline package.json]: I think we should make sure to clean out the repository before we run the build, so we don't by accident rely on old files from an earlier build

File: gulpfile.js (hunk 10 above)
Comment (mlunoe) [inline gulpfile.js]: Sounds good

File: gulpfile.js (hunk 12 above)
Comment (MatApple) [inline gulpfile.js]: Sorry, yes the config file was removed. There was so much config in it that isn't being used anymore

Comment (MatApple) [inline gulpfile.js]: Fixed

File: gulpfile.js (hunk 9 above)
Comment (mlunoe) [inline gulpfile.js]: Ok, maybe this could also be marked as an improvement, but I think the webpack server could take care of this too, right?

Comment (MatApple) [inline gulpfile.js]: Yeah it can. Its just a little more involved because we need to write a plugin.

File: package.json (hunk 11 above)
Comment (MatApple) [inline package.json]: OK will update to clean out `dist`

Comment (MatApple) [inline package.json]: Done

File: gulpfile.js (hunk 9 above)
Comment (mlunoe) [inline gulpfile.js]: Ok, let's mark it as an improvement then!

File: webpack/webpack.config.babel.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+// Can override this with npm config set dcos-ui:external_plugins ../some/relative/path/to/repo
+let pluginsDir = absPath(process.env.npm_package_config_external_plugins || 'plugins');
```
Comment (rcorral) [inline webpack/webpack.config.babel.js]: What's the purpose of defaulting to `'plugins'`?

File: gulpfile.js [hunk 14]
```diff
@@ -1,261 +1,18 @@
...
-  }
-}
 
-function ensurePluginDirectoryExists() {
-  // Make temp plugins directory if doesn't exist
-  if (!fs.existsSync(config.dirs.pluginsTmp)) {
-    mkdirp(config.dirs.pluginsTmp, function () {
-      console.log('Created ' + config.dirs.pluginsTmp);
-    });
+gulp.task('ensureConfig', function () {
```
Comment (rcorral) [inline gulpfile.js]: I'm not sure this is needed. Why can't the build just fail until someone creates the config.dev file?

Comment (MatApple) [inline gulpfile.js]: It will break CI

File: README.md (hunk 1 above)
Comment (rcorral) [inline README.md]: Can this be removed?

File: jest/genConfig.js [hunk 15]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (rcorral) [inline jest/genConfig.js]: Could we make this filename `gen-config.js` I don't think it needs to be camel cased.

File: webpack/webpack.config.babel.js (hunk 13 above)
Comment (MatApple) [inline webpack/webpack.config.babel.js]: Defaulting to `plugins` provides us with a resolvable directory so we don't get any errors. It won't be used other than to safely resolve `require.context` in `Loader.js`. If pluginsDir is invalid (can only happen if a developer does `npm config set dcos-ui:external_plugins some/dir`) where `some/dir` is invalid then we throw an error because it's not resolvable.

File: package.json [hunk 16]
```diff
@@ -47,121 +52,65 @@
...
     "babel-runtime": "6.6.1",
-    "browser-sync": "2.12.3",
-    "del": "2.2.0",
-    "envify": "3.4.0",
+    "css-loader": "0.23.1",
     "eslint": "2.0.0",
     "eslint-loader": "1.3.0",
     "eslint-plugin-react": "2.6.4",
+    "extract-text-webpack-plugin": "1.0.1",
+    "file-loader": "0.8.5",
```
Comment (rcorral) [inline package.json]: Where is this used?

File: package.json [hunk 17]
```diff
@@ -47,121 +52,65 @@
...
     "babel-preset-es2015": "6.6.0",
+    "babel-preset-jest": "11.0.2",
     "babel-preset-react": "6.5.0",
+    "babel-preset-react-hmre": "1.1.1",
+    "babel-register": "6.8.0",
     "babel-runtime": "6.6.1",
-    "browser-sync": "2.12.3",
-    "del": "2.2.0",
-    "envify": "3.4.0",
+    "css-loader": "0.23.1",
```
Comment (rcorral) [inline package.json]: Is this needed for `postcss-loader`?

File: README.md (hunk 1 above)
Comment (MatApple) [inline README.md]: Yes, I still need to update the README to reflect the new dev and build process.

File: package.json [hunk 18]
```diff
@@ -47,121 +52,65 @@
...
     "webpack": "1.13.0",
+    "webpack-dev-server": "1.14.1",
     "webpack-notifier": "1.3.0"
   },
   "scripts": {
-    "clean": "rm -rf build",
+    "clean": "rm -rf dist",
     "prebuild-ee": "./scripts/validate-tests",
-    "build-ee": "npm run clean && npm test && NODE_ENV='production' ./node_modules/.bin/gulp dist",
+    "build-ee": "./node_modules/.bin/gulp ensureConfig && npm run clean && npm run lint && npm run lint-plugins && npm test && npm run build",
```
Comment (rcorral) [inline package.json]: Can we rename this to `build-with-plugins`

File: jest/genConfig.js (hunk 15 above)
Comment (MatApple) [inline jest/genConfig.js]: Done

File: package.json [hunk 19]
```diff
@@ -47,121 +52,65 @@
...
     "dist": "npm run build-ee",
     "prebuild-ce": "./scripts/validate-tests",
-    "build-ce": "npm run clean && npm test && NODE_ENV='production' ./node_modules/.bin/gulp dist",
-    "serve": "NODE_ENV='development' ./node_modules/.bin/gulp serve",
+    "build-ce": "./node_modules/.bin/gulp ensureConfig && npm run clean && npm run lint && npm test && npm run build",
```
Comment (rcorral) [inline package.json]: Maybe this can be `build-without-plugins`

File: package.json (hunk 16 above)
Comment (MatApple) [inline package.json]: To load images :)

File: package.json (hunk 17 above)
Comment (MatApple) [inline package.json]: `postcss-loader` basically provides you with entry points to run `autoprefixer` etc. Then to load the result we still need `css-loader`.

File: webpack/webpack.config.babel.js (hunk 13 above)
Comment (rcorral) [inline webpack/webpack.config.babel.js]: Shouldn't this variable be called `externalPluginsDir`?

Comment (rcorral) [inline webpack/webpack.config.babel.js]: I think it's ok for the build to fail. The goal is to build the application the way you have your config set rather than building the application in a different way that what you'd want.

File: webpack/webpack.dev.babel.js [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+// Defaults to value in package.json.
+// Can override with npm config set dcos-ui:port 80
+let PORT = process.env.npm_package_config_port;
+let proxy;
+
+try {
+  // If this fails - you should create a proxy.dev.js by copying proxy.template.js
+  // and customizing the proxy as needed.
+  proxy = require('./proxy.dev.js');
```
Comment (rcorral) [inline webpack/webpack.dev.babel.js]: Again here, I think it's ok to fail if the file doesn't exist.
Maybe we can have an npm script that just copies this proxy file to `.dev` and same thing for the `config.template` that way it's easy for someone to get started.

Comment (MatApple) [inline webpack/webpack.dev.babel.js]: Good idea, but I wouldn't want someone to start the server for the first time and run into an error. I'll create script.

File: gulpfile.js (hunk 14 above)
Comment (rcorral) [inline gulpfile.js]: We can add a line to CI to create the file before the tests run.

File: webpack/webpack.dev.babel.js (hunk 20 above)
Comment (MatApple) [inline webpack/webpack.dev.babel.js]: Done

File: package.json (hunk 19 above)
Comment (MatApple) [inline package.json]: It's now just `build` and `build-with-plugins`

File: gulpfile.js (hunk 14 above)
Comment (MatApple) [inline gulpfile.js]: Thats fair, but I think since we have it, it doesn't make sense to remove it since its helping people get up and running quicker. It's nice for someone to clone the repo and `npm i && npm start`. In exchange for removing this, we'd have to write more docs (which we need at some point, just not immediately) :). What do you think?

File: webpack/webpack.config.babel.js (hunk 13 above)
Comment (MatApple) [inline webpack/webpack.config.babel.js]: Could you suggest a default configuration so it won't fail to build by default? Most people who get started won't have an external plugins directory. I updated the name of the variable to be more explicit!

File: package.json [hunk 21]
```diff
@@ -47,121 +52,64 @@
...
-    "test": "./node_modules/.bin/gulp copy:external-plugins && npm run lint && jest --no-cache"
+    "lint-plugins": "PLUGINS=$npm_package_config_external_plugins; ./node_modules/.bin/eslint \"$PLUGINS/**/*.js\" --quiet",
+    "prebuild": "./scripts/validate-tests",
+    "serve": "NODE_ENV='development' npm run dist && ./node_modules/.bin/gulp serve",
+    "serve-notify": "export NOTIFY=true && npm run serve",
+    "scaffold": "./node_modules/.bin/gulp ensureConfig && ./node_modules/.bin/gulp ensureDevProxy",
+    "start": "PORT=$npm_package_config_port; webpack-dev-server --config ./webpack/webpack.dev.babel.js --hot --progress --colors --content-base ./build --port $PORT",
+    "test": "node jest/gen-config.js && NODE_PATH=node_modules jest --config=jest/config.json --no-cache",
+    "testing": "NODE_ENV='testing' npm run dist && ./node_modules/.bin/gulp serve",
+    "dist": "NODE_ENV='production' webpack --config ./webpack/webpack.production.babel.js"
```
Comment (rcorral) [inline package.json]: I like the idea of renaming `dist` to something else.
What if this became `build-application`?

File: README.md [hunk 22]
```diff
@@ -36,8 +36,6 @@ Node 4.x (and above) is **required**. We suggest using [nvm](https://github.com/
 6. Navigate to http://localhost:4200
 
 
-After installing all development dependencies and configuring your local environment (steps 1-4 above), you can run test by running `npm test`.  You can build production assets by running `npm run dist`.  If you are actively developing, you may want to run `npm start` to avoid having to refresh your browser with every change.
-
```
Comment (MatApple) [inline README.md]: This info is not needed anymore as its explained in previous steps or further down the page.
